### PR TITLE
[mlir][openmp] - Add the depend clause to omp.target and related offloading directives

### PR DESCRIPTION
**Pull request by bhandarkar-pranav**
This patch adds support for the depend clause in a number of OpenMP directives/constructs related to offloading. Specifically, it adds the handling of the depend clause when it is used with the following constructs

- target
- target enter data
- target update data
- target exit data

### CODE REVIEW

**Comment by bhandarkar-pranav** [comment]
@skatrak @clementval @kiranchandramohan - PTAL.
As per @clementval's request, I have closed my previous PR (https://github.com/llvm/llvm-project/pull/80626) in order to split that into two PRs, one for the MLIR related changes (this one) and one for the flang related changes only (subsequent PR to follow)

**Comment by clementval** [comment]
Thanks for splitting it. If possible we tend to always split MLIR core changes with Flang lowering changes.

**Review by skatrak — APPROVED** [review]
Thanks for this work, LGTM. Give it a day or so before merging in case others have any concerns.

**Comment by kiranchandramohan** [comment]
Thanks for the update. Could you fix the CI failure by passing empty values in Flang for the modified operations?

I believe we do not support array dependencies. I guess it might be possible to use `map_info` or similar to add support.

**Comment by bhandarkar-pranav** [comment]
> I believe we do not support array dependencies. I guess it might be possible to use `map_info` or similar to add support.
Hi @kiranchandramohan  - I am sorry I don't understand what you mean by not supporting array dependencies? You mean in the codegen to LLVMIR from FIR?

**Comment by kiranchandramohan** [comment]
> > I believe we do not support array dependencies. I guess it might be possible to use `map_info` or similar to add support.
> 
> Hi @kiranchandramohan - I am sorry I don't understand what you mean by not supporting array dependencies? You mean in the codegen to LLVMIR from FIR?

This was just a side comment and not related to the patch. I was saying that we do not support task dependence with array sections as arguments. They will hit the TODO in flang lowering. https://github.com/llvm/llvm-project/blob/a0635edc5980218ad210da25a5c9afe346110ccb/flang/lib/Lower/OpenMP.cpp#L1777
Also the codegen to LLVM IR is also not support.

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Comment by bhandarkar-pranav** [comment]
@kiranchandramohan @clementval @skatrak  Is this ok to merge now?  The CI failure (windows64) doesn't look related to me?

**Comment by bhandarkar-pranav** [comment]
> @kiranchandramohan @clementval @skatrak Is this ok to merge now? The CI failure (windows64) doesn't look related to me?

Can one of you merge this for me, I do not have write access.

**Pull request merged**
